### ISSUE TITLE
DSA support

### ISSUE THREAD

**Issue by blogic** [mstpd/mstpd#111]
Hi,
testing the daemon with OpenWrt HEAD on Realtek switches. I can see the BDPU TX go out and can sniff it on an externally attached PC. If i loop 2 ports, the RX code receives a packet, but ifindex is that of the bridge and not of the port. The code cycles through all registered bridges and checks ifindex vs the member ports ifindex. This however fails due to described reason and the packet is discarded.
I have not dug deeper into this but my assumption is that possibly the kernel or the bpf behaves differently on a bridge with switchdev interface members as opposed to normal netdev members causing the ifindex to be something we do not expect.
so before draining lots of time into this, is this code base known to be working when using DSA or not or we do not know ?!
Thanks in advance!
   John

**Comment by commodo** [comment]
maybe @vivien can share some input on the DSA stuff;
i remember talking to him and he was trying out some things; but i am not sure where he got to;

what i can tell you is that i got this working with swconfig [libswconfig] on OpenWrt with the b53 driver [from OpenWrt], back in ~2015 ; this was on some mpc85xx variant (derived from TL-WDR4900)

in fact [all the procd & UCI in here](https://github.com/openwrt/packages/blob/master/net/mstpd/files/etc/init.d/mstpd.init) was working with the combination described above;
this was when switchdev and DSA were appearing around, nobody knew which was going to be the one that dominates [because they were both evolving in parallel];
[a lot of the patches i sent to OpenWrt](https://patchwork.ozlabs.org/project/openwrt/list/?series=&submitter=65503&state=*&q=&archive=&delegate=) back then were to support STP & LLDP with b53 [including ARL manipulation and so on]

anyway, what we did [with b53] was:
* `eth0` was the CPU port moving all the relevant to the host system
* we created these `lanX` [0 to 7] interfaces for each physical port; the b53 switch chip was configured to forward BPDUs and LLPD frames to these ports; mstpd and lldpd would work with the data on these ports
  * for lldpd, just using these interfaces was enough
  * for mstpd we had to do some extra modifications [as you can guess]

For the modifications for `mstpd`:
1. We implemented our own hooks in `drivers_deps.c` which would set some STP states in the b53 chip
2. We did a few modifications inside the `bridge_track.c` [and probably other places, but I remember this to be the main place of modifications]
   * we had to hijack a lot of the ifindex code [as you've noticed], and the way we did it was with nasty `#ifdef LIBSWCONFIG` compile flags
   * naturally, in this libswconfig implementation [and I suspect DSA as well], a lot of the information in `/sys/class/net` was useless and would only cause issues; we were using the port IDs from libswconfig
   * I did have the idea to create a sort of `libsw/libbridge` library out of this `bridge_track.c`, one that exports an API that is agnostic of libswconfig or Linux bridges, but never had time; the source code is in some repo at some company, because it was bad enough, I never bothered to send it upstream [i.e.the libswconfig patches]

We found that the RSTP implementation is compatible with all other switches [Cisco, HP, etc] so this worked well in all production cases. For MSTP, unfortunately some HP switches did not like the MSTP implementation, so... yeah (more work is needed somewhere). The decision was that RSTP is good enough.

Oh, the current implementation of mstpd with [the UCI + procd stuff](https://github.com/openwrt/packages/blob/master/net/mstpd/files/etc/init.d/mstpd.init) is known to work with OpenWrt + Linux bridges. I don't know if this is still the case today, but we had this powerful GW product (running OpenWrt on a PPC64) that had individual ethernet ports [no switch-chip] and it required to be part of an STP topology.
So, the `/sys/class/net` works only with standard Linux bridges.

On an unrelated note, I just remembered I have an WRT1200AC around that I bought recently, and that seems to have some Marvell chip with driver support in Linux DSA, so I may try it [not sure when though].

**Comment by pavlix** [comment]
Good to know. I'm experimenting with a fully offloaded MSTP setup. I described my OpenWRT/UCI ideas in another issue: 

https://github.com/mstpd/mstpd/pull/103#issuecomment-798801923

**Comment by commodo** [comment]
> Good to know. I'm experimenting with a fully offloaded MSTP setup. I described my OpenWRT/UCI ideas in another issue:
> 
> [#103 (comment)](https://github.com/mstpd/mstpd/pull/103#issuecomment-798801923)

I'd say that first thing for mstpd, would be to rework `bridge_track.c` and maybe `driver_deps.c` into a `libsw` that can be somehow swapped under the hood.

**Comment by pavlix** [comment]
> I'd say that first thing for mstpd, would be to rework `bridge_track.c` and maybe `driver_deps.c` into a `libsw` that can be somehow swapped under the hood.

I've done it for the push operations (i.e. not for monitoring messages from kernel yet) in #112 and this could later be extended to handle the monitor cases as well.

**Comment by Bartel-C8** [comment]
@pavlix What is the status of using MSTPD in combination with a DSA driver? Is it possible to use just RSTP already e.g.?

**Comment by commodo** [comment]
> > > @commodo , that DSA restriction, does it also apply to normal RSTP without using VLANs? Thanks!
> > 
> > 
> > so, DSA is not implemented yet at all in master; there's some effort for getting there, but it's slow;
> 
> Is there a branch / fork where that development is being done?

There's a few PRs that you can take a look at.
This one looks like a good start: https://github.com/mstpd/mstpd/pull/103

**Comment by pavlix** [comment]
> Hello,@pavlix What is the status of using MSTPD in combination with a DSA driver? how to support mstp or how to config to make mstp work on the DSA switch?

Hello! I can certainly help with reviews but I'm not often looking at the project on GitHub. I experimented with that and there are pull requests including kernel changes to make it work. It was okay but the kernel changes were needed in order to pass the right data structures down the driver.

**Comment by Bartel-C8** [comment]
In the meantime I started using this RSTP daemon, from OpenWRT. Which does work with DSA.
https://github.com/openwrt/ustp

It currently has a (soft) dependency, you get warnings, on a out-of-tree (rejected) kernel patch: https://github.com/openwrt/openwrt/blob/master/target/linux/generic/pending-5.10/710-bridge-add-knob-for-filtering-rx-tx-BPDU-pack.patch
https://lore.kernel.org/all/d4f1f9b1-6e8e-d21d-603f-7a0889e33a78@nvidia.com/

But even without the patch, it works great. I have it working on a custom buildroot build.

**Comment by commodo** [comment]
> As it is inspired/based on this project (I guess)? Maybe some inspiration could be fetched to unify this?

That could work somehow.
We could also have an OpenWrt build-flag where (when enabled) we compile in the libubox dependencies.
That is if @nbd168 does not mind also.

**Comment by KanjiMonster** [comment]
I created a proof of concept implementation using kernel MST support. I tested it with local changes adding MST support to b53 (where most of the chips support it).
